src/lib/block_producer/block_producer.ml

```diff
@@ -656,441 +547,559 @@ let validate_genesis_protocol_state_block ~genesis_state_hash (b, v) =
+          ]
+        in
+        [%log internal] "Wait_for_confirmation" ;
+        [%log debug] ~metadata
+          "Waiting for block $state_hash to be inserted into frontier" ;
+        Deferred.choose
+          [ Deferred.choice
+              (Transition_registry.register transition_registry
+                 protocol_state_hashes.state_hash )
+              (Fn.const (Ok `Transition_accepted))
+          ; Deferred.choice
+              ( Block_time.Timeout.create time_controller
+                  (* We allow up to 20 seconds for the transition
+                     to make its way from the transition_writer to
+                     the frontier.
+                     This value is chosen to be reasonably
+                     generous. In theory, this should not take
+                     terribly long. But long cycles do happen in
+                     our system, and with medium curves those long
+                     cycles can be substantial.
+                  *)
+                  (Block_time.Span.of_ms 20000L)
+                  ~f:(Fn.const ())
+              |> Block_time.Timeout.to_deferred )
+              (Fn.const (Ok `Timed_out))
+          ]
+        >>= function
+        | `Transition_accepted ->
+            [%log internal] "Transition_accepted" ;
+            [%log info] ~metadata
+              "Generated transition $state_hash was accepted into transition \
+               frontier" ;
+            Deferred.map ~f:Result.return
+              (Mina_networking.broadcast_state net
+                 (Breadcrumb.block_with_hash breadcrumb) )
+        | `Timed_out ->
+            (* FIXME #3167: this should be fatal, and more
+               importantly, shouldn't happen.
+            *)
+            [%log internal] "Transition_accept_timeout" ;
+            let msg : (_, unit, string, unit) format4 =
+              "Timed out waiting for generated transition $state_hash to enter \
+               transition frontier. Continuing to produce new blocks anyway. \
+               This may mean your CPU is overloaded. Consider disabling \
+               `-run-snark-worker` if it's configured."
+            in
+            let span = Block_time.diff (Block_time.now time_controller) start in
+            let metadata =
+              [ ("time", `Int (Block_time.Span.to_ms span |> Int64.to_int_exn))
+              ; ("protocol_state", Protocol_state.Value.to_yojson protocol_state)
+              ]
+              @ metadata
+            in
+            [%log' debug rejected_blocks_logger] ~metadata msg ;
+            [%log fatal] ~metadata msg ;
+            return ()
+      in
+      let%bind res = emit_breadcrumb () in
+      let span = Block_time.diff (Block_time.now time_controller) start in
+      handle_block_production_errors ~logger ~rejected_blocks_logger
+        ~time_taken:span ~previous_protocol_state ~protocol_state res
+
+let iteration ~schedule_next_check ~dispatch_now ~schedule_dispatch
```

Comment:
Rename:

* `~schedule_next_check` to `~schedule_next_vrf_check`
* `~dispatch_now` to `~produce_block_now`
* `~next_check_now` to `~next_vrf_check_now`
* `~schedule_dispatch` to `~schedule_block_production`